vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_gpio.c

```diff
@@ -0,0 +1,622 @@
+    {
+        /* Update was successful, update descriptor to reflect new settings */
+        pxGpio->xConfig = *pxNewConfig;
+    }
+
+    return lReturnCode;
+}
+
+/*---------------------------------------------------------------------------------*
+*                               API Implementation                                *
+*---------------------------------------------------------------------------------*/
+
+IotGpioHandle_t iot_gpio_open( int32_t lGpioNumber )
+{
+    if( !prvIsValidPinIndex( lGpioNumber ) )
+    {
+        IOT_GPIO_LOGF( " Incorrect pin index[%d]. Please verify IOT_COMMON_IO_GPIO_PIN_MAP\r\n",
+                       lGpioNumber );
+        return NULL;
+    }
+
+    if( pxGpioDesc[ lGpioNumber ].ucState != IOT_GPIO_CLOSED )
+    {
+        return NULL;
+    }
+
+    /* Configure pin to default settings. */
+    IotGpioHandle_t pxGpio = &pxGpioDesc[ lGpioNumber ];
+    *pxGpio = ( IotGpioDescriptor_t ) DEFAULT_GPIO_DESCRIPTOR;
+
+    if( IOT_GPIO_SUCCESS == prvConfigurePin( pxGpio, &pxGpio->xConfig ) )
+    {
+        /* Claim descriptor */
+        pxGpio->lGpioNumber = lGpioNumber;
+        pxGpio->ucState = IOT_GPIO_OPENED;
+
+        return pxGpio;
+    }
+
+    return NULL;
+}
+
+void iot_gpio_set_callback( IotGpioHandle_t const pxGpio,
+                            IotGpioCallback_t xGpioCallback,
+                            void * pvUserContext )
+{
+    if( !prvIsValidHandle( pxGpio ) || ( xGpioCallback == NULL ) )
+    {
+        return;
+    }
+
+    pxGpio->xUserCallback = xGpioCallback;
+    pxGpio->pvUserContext = pvUserContext;
+}
+
+
+int32_t iot_gpio_read_sync( IotGpioHandle_t const pxGpio,
+                            uint8_t * pucPinState )
+{
+    if( !prvIsValidHandle( pxGpio ) || ( pxGpio->xConfig.xDirection != eGpioDirectionInput ) )
+    {
+        return IOT_GPIO_INVALID_VALUE;
+    }
+
+    *pucPinState = nrf_gpio_pin_read( plGpioMap[ pxGpio->lGpioNumber ] );
+
+    return IOT_GPIO_SUCCESS;
+}
+
+
+int32_t iot_gpio_write_sync( IotGpioHandle_t const pxGpio,
+                             uint8_t ucPinState )
+{
+    if( !prvIsValidHandle( pxGpio ) || ( pxGpio->xConfig.xDirection != eGpioDirectionOutput ) )
+    {
+        return IOT_GPIO_INVALID_VALUE;
+    }
+
+    nrf_gpio_pin_write( plGpioMap[ pxGpio->lGpioNumber ], ucPinState );
+
+    return IOT_GPIO_SUCCESS;
+}
+
+int32_t iot_gpio_close( IotGpioHandle_t const pxGpio )
+{
+    if( !prvIsValidHandle( pxGpio ) )
+    {
+        return IOT_GPIO_INVALID_VALUE;
+    }
+
+    /* De-init, setting pin to board's default settings for unused pins, then unclaim*/
+    uint32_t ulPin_NRF = plGpioMap[ pxGpio->lGpioNumber ];
+
+    if( ( pxGpio->xConfig.xDirection == eGpioDirectionInput ) &&
+        ( pxGpio->xConfig.xInterruptMode != eGpioInterruptNone ) )
+    {
+        /* GPIOTE require separate deinit */
+        nrf_drv_gpiote_in_uninit( ulPin_NRF );
+    }
+
+    nrf_gpio_cfg_default( ulPin_NRF );
+    pxGpio->ucState = IOT_GPIO_CLOSED;
```

Comment:
Reset the GPIO config back to default value on close ?

Reply:
Yes, the default is high-impedance input pin. This was mostly done for "baby-sitting" purposes.

This way if someone was maniacally connecting closed GPIOs at random, they wouldn't accidentally short-circuit from residual OUT-HIGH an OUT-LOW.

Reply:
Descriptor is initialized upon `iot_gpio_open`, but in the latest commit I added code to reconfigure defaults at close as well.